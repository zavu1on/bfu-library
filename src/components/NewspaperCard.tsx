import { FC } from 'react'
import { Col } from 'react-bootstrap'
import { Link } from 'react-router-dom'
import star from '../static/star-fill.svg'

interface IProps {
  id: number | string
  imageUrl: string
  alt: string
  date: number | string | undefined
  tags?: {
    name: string
  }[]
  size: number
  link: string
  tagsClassName?: string
  isFavorite?: boolean
  favoriteClickHandler?: any
}

export const NewspaperCard: FC<IProps> = ({
  id,
  imageUrl,
  alt,
  date,
  size,
  link,
  tags,
  tagsClassName,
  isFavorite,
  favoriteClickHandler,
}) => {
  if (tags?.length) {
    return (
      <Col sm={size}>
        <div className='publisher-year-col margin-on-phone'>
          <Link to={link}>
            <img src={imageUrl} alt={alt} className='squared-img' />
            <div className='text'>
              <div>{date}</div>
              {isFavorite ? (
                <button
                  className='star'
                  onClick={event => {
                    event.preventDefault()

                    favoriteClickHandler()
                  }}
                >
                  <img src={star} alt='star' />
                </button>
              ) : null}
            </div>
          </Link>
        </div>
        <div className={'tags-container ' + tagsClassName}>
          {tags.map(t => (
            <div className='tags-item' key={t.name}>
              <span>{t.name}</span>
            </div>
          ))}
        </div>
      </Col>
    )
  }

  return (
    <Col key={id} sm={size}>
      <div className='publisher-year-col margin-on-phone'>
        <Link to={link}>
          <img src={imageUrl} alt={alt} className='squared-img' />
          <div className='text'>
            <div>{date}</div>
            {isFavorite ? (
              <button
                className='star'
                onClick={event => {
                  event.preventDefault()

                  favoriteClickHandler()
                }}
              >
                <img src={star} alt='star' />
              </button>
            ) : null}
          </div>
        </Link>
      </div>
    </Col>
  )
}
