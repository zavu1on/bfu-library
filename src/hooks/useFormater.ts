export const useFormater = () => {
  return function (date: Date | string | undefined): string | undefined {
    if (typeof date === 'string' || typeof date === 'undefined') {
      return date
    }

    let m = date.getMonth()
    let year = date.getFullYear()
    const d = date.getDate()

    if (m === 0) {
      m = 12
      year -= 1
    }

    return `${year}-${m >= 10 ? m : '0' + m}-${d >= 10 ? d : '0' + d}`
  }
}
