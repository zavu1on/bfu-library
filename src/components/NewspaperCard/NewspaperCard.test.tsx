import { NewspaperCard } from '../NewspaperCard'
import { render, screen, fireEvent } from '@testing-library/react'
import jest from 'jest'
import { BrowserRouter, Route, Routes } from 'react-router-dom'
import { useFormater } from '../../hooks/useFormater/useFormater'

const _ = useFormater()

describe('NewspaperCard', () => {
  test('Карточка без тегов', () => {
    const { container } = render(
      <BrowserRouter>
        <Routes>
          <Route
            path='/'
            element={
              <NewspaperCard
                id={1}
                imageUrl={'https://place-hold.it/350x480/'}
                alt={'test image'}
                date={_(new Date())}
                size={3}
                link={`/archive/newspapers/1/`}
              />
            }
          />
        </Routes>
      </BrowserRouter>
    )

    screen.getByText(_(new Date()))

    const image = screen.getByAltText('test image') as HTMLImageElement
    expect(image.src).toEqual('https://place-hold.it/350x480/')

    expect(!!container.querySelector('.col-sm-3')).toBeTruthy()
  })

  test('Карточка с тегами', () => {
    const { container } = render(
      <BrowserRouter>
        <Routes>
          <Route
            path='/'
            element={
              <NewspaperCard
                id={1}
                imageUrl={'https://place-hold.it/350x480/'}
                alt={'test image'}
                date={_(new Date())}
                size={3}
                link={`/archive/newspapers/1/`}
                tagsClassName='test-class-name'
                tags={[
                  {
                    name: 'Кино',
                  },
                  {
                    name: 'Игры',
                  },
                  {
                    name: 'Фантастика',
                  },
                  {
                    name: 'Тренд',
                  },
                ]}
              />
            }
          />
        </Routes>
      </BrowserRouter>
    )

    screen.getByText(_(new Date()))

    const image = screen.getByAltText('test image') as HTMLImageElement
    expect(image.src).toEqual('https://place-hold.it/350x480/')

    expect(!!container.querySelector('.col-sm-3')).toBeTruthy()

    const tags = container.querySelectorAll(
      '.tags-container.test-class-name .tags-item'
    )
    expect(tags.length).toEqual(4)
    expect(tags).toMatchSnapshot()
  })

  test('Карточка без тегов, отмечена избранной', () => {
    let hasClicked = false
    const setHasClicked = () => (hasClicked = true)

    const { container } = render(
      <BrowserRouter>
        <Routes>
          <Route
            path='/'
            element={
              <NewspaperCard
                id={1}
                imageUrl={'https://place-hold.it/350x480/'}
                alt={'test image'}
                date={_(new Date())}
                size={3}
                link={`/archive/newspapers/1/`}
                isFavorite={true}
                favoriteClickHandler={setHasClicked}
              />
            }
          />
        </Routes>
      </BrowserRouter>
    )

    screen.getByText(_(new Date()))

    const image = screen.getByAltText('test image') as HTMLImageElement
    expect(image.src).toEqual('https://place-hold.it/350x480/')

    expect(!!container.querySelector('.col-sm-3')).toBeTruthy()
    expect(!!screen.getByAltText('star')).toBeTruthy()

    const btn = container.querySelector('button.star')
    expect(!!btn).toBeTruthy()
    fireEvent.click(btn!)
    expect(hasClicked).toBeTruthy()
  })

  test('Карточка с тегами, отмечена избранной', () => {
    let hasClicked = false
    const setHasClicked = () => (hasClicked = true)

    const { container } = render(
      <BrowserRouter>
        <Routes>
          <Route
            path='/'
            element={
              <NewspaperCard
                id={1}
                imageUrl={'https://place-hold.it/350x480/'}
                alt={'test image'}
                date={_(new Date())}
                size={3}
                link={`/archive/newspapers/1/`}
                tagsClassName='test-class-name'
                tags={[
                  {
                    name: 'Кино',
                  },
                  {
                    name: 'Игры',
                  },
                  {
                    name: 'Фантастика',
                  },
                  {
                    name: 'Тренд',
                  },
                ]}
                isFavorite={true}
                favoriteClickHandler={setHasClicked}
              />
            }
          />
        </Routes>
      </BrowserRouter>
    )

    screen.getByText(_(new Date()))

    const image = screen.getByAltText('test image') as HTMLImageElement
    expect(image.src).toEqual('https://place-hold.it/350x480/')

    const tags = container.querySelectorAll(
      '.tags-container.test-class-name .tags-item'
    )
    expect(tags.length).toEqual(4)
    expect(tags).toMatchSnapshot()

    expect(!!container.querySelector('.col-sm-3')).toBeTruthy()
    expect(!!screen.getByAltText('star')).toBeTruthy()

    const btn = container.querySelector('button.star')
    expect(!!btn).toBeTruthy()
    fireEvent.click(btn!)
    expect(hasClicked).toBeTruthy()
  })
})
