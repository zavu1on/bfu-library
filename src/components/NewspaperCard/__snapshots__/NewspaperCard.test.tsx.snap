// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewspaperCard Карточка с тегами 1`] = `
NodeList [
  <div
    class="tags-item"
  >
    <span>
      Кино
    </span>
  </div>,
  <div
    class="tags-item"
  >
    <span>
      Игры
    </span>
  </div>,
  <div
    class="tags-item"
  >
    <span>
      Фантастика
    </span>
  </div>,
  <div
    class="tags-item"
  >
    <span>
      Тренд
    </span>
  </div>,
]
`;

exports[`NewspaperCard Карточка с тегами, отмечена избранной 1`] = `
NodeList [
  <div
    class="tags-item"
  >
    <span>
      Кино
    </span>
  </div>,
  <div
    class="tags-item"
  >
    <span>
      Игры
    </span>
  </div>,
  <div
    class="tags-item"
  >
    <span>
      Фантастика
    </span>
  </div>,
  <div
    class="tags-item"
  >
    <span>
      Тренд
    </span>
  </div>,
]
`;
